test/relay/auto-relay.node.js

```diff
@@ -452,4 +460,125 @@ describe('auto-relay', () => {
+              hop: {
+                enabled: index === 2
+              },
+              autoRelay: {
+                enabled: true,
+                maxListeners: 1
+              }
+            }
+          }
+        }
+
+        return new Libp2p({
+          ...opts,
+          modules: {
+            ...opts.modules,
+            contentRouting: [delegate]
+          },
+          addresses: {
+            listen: [listenAddr]
+          },
+          connectionManager: {
+            autoDial: false
+          },
+          peerDiscovery: {
+            autoDial: false
+          },
+          peerId
+        })
+      })
+
+      sinon.spy(relayLibp2p.contentRouting, 'provide')
+    })
+
+    beforeEach(async () => {
+      nock('http://0.0.0.0:60197')
+        // mock the refs call
+        .post('/api/v0/refs')
+        .query(true)
+        .reply(200, null, [
+          'Content-Type', 'application/json',
+          'X-Chunked-Output', '1'
+        ])
+
+      // Start each node
+      await Promise.all([libp2p, libp2p2, relayLibp2p].map(libp2p => libp2p.start()))
+
+      // Should provide on start
+      await pWaitFor(() => relayLibp2p.contentRouting.provide.callCount === 1)
+
+      const provider = relayLibp2p.peerId.toB58String()
+      const multiaddrs = relayLibp2p.multiaddrs.map((m) => m.toString())
+
+      // Mock findProviders
+      nock('http://0.0.0.0:60197')
+        .post('/api/v0/dht/findprovs')
+        .query(true)
+        .reply(200, `{"Extra":"","ID":"${provider}","Responses":[{"Addrs":${JSON.stringify(multiaddrs)},"ID":"${provider}"}],"Type":4}\n`, [
+          'Content-Type', 'application/json',
+          'X-Chunked-Output', '1'
+        ])
+    })
+
+    afterEach(() => {
+      // Stop each node
+      return Promise.all([libp2p, libp2p2, relayLibp2p].map(libp2p => libp2p.stop()))
+    })
+
+    it('should find providers for relay and add it as listen relay', async () => {
+      const originalMultiaddrsLength = libp2p.multiaddrs.length
+
+      // Spy add listen relay
+      sinon.spy(libp2p.relay._autoRelay, '_addListenRelay')
+      // Spy Find Providers
+      sinon.spy(libp2p.contentRouting, 'findProviders')
+
+      // Try to listen on Available hop relays
+      await libp2p.relay._autoRelay._listenOnAvailableHopRelays()
+
+      // Should try to find relay service providers
+      await pWaitFor(() => libp2p.contentRouting.findProviders.callCount === 1)
+      // Wait for peer added as listen relay
+      await pWaitFor(() => libp2p.relay._autoRelay._addListenRelay.callCount === 1)
+      expect(libp2p.relay._autoRelay._listenRelays.size).to.equal(1)
+      await pWaitFor(() => libp2p.multiaddrs.length === originalMultiaddrsLength + 1)
+
+      const relayedAddr = libp2p.multiaddrs[libp2p.multiaddrs.length - 1]
+      libp2p2.peerStore.addressBook.set(libp2p2.peerId, [relayedAddr])
+
+      // Dial from peer 2 through the relayed address
+      const conn = await libp2p2.dial(libp2p2.peerId)
```

Comment:
This is dialing itself, it should be failing. 

Also, I'd recommend changing the names of these nodes because `libp2p2` and `libp2p` is difficult to quickly distinguish. (local/remote, source/dest, etc)

Reply:
> This is dialing itself, it should be failing.

lol yes, it works because of the mock response. I will change the names as you suggest